src/server/tran_server.cpp

```diff
@@ -362,75 +370,60 @@ void
 tran_server::disconnect_all_page_servers ()
 {
   assert_is_tran_server ();
+  for (auto &conn : m_page_server_conn_vec)
+    {
+      constexpr bool with_disconnect_msg = true;
+      conn->disconnect_async (with_disconnect_msg);
+    }
 
-  // We assumes that no threads are waiting for response at this moment so that m_conn->stop_response_broker() is not needed.
-  std::vector<std::unique_ptr<connection_handler>> conn_vec;
-  {
-    std::lock_guard<std::shared_mutex> lk_guard (m_page_server_conn_vec_mtx);
-    m_page_server_conn_vec.swap (conn_vec);
-  }
-  // finalize connections out of the mutex, m_page_server_conn_vec_mtx, since it joins request handler thread internally, and receive_disconnect_request() also acquires the lock of the mutex.
-  conn_vec.clear ();
-  m_main_conn_cv.notify_all ();
-
-  // Wait until all disconnection reuqests are handled.
-  m_async_disconnect_handler.terminate ();
+  for (auto &conn : m_page_server_conn_vec)
+    {
+      conn->wait_async_disconnection ();
+    }
 
   er_log_debug (ARG_FILE_LINE, "Transaction server disconnected from all page servers.");
 }
 
-void
-tran_server::disconnect_page_server_async (const connection_handler *conn)
+int
+tran_server::reset_main_connection ()
 {
-  bool main_conn_changed = false;
-  std::string prev_main_conn_id;
   auto &conn_vec = m_page_server_conn_vec;
-  auto ulock = std::unique_lock<std::shared_mutex> (m_page_server_conn_vec_mtx);
-
-  auto conn_it = std::find_if (conn_vec.begin (), conn_vec.end (),
-			       [&conn] (std::unique_ptr<connection_handler> &conn_uptr)
+  /* the priority to select the main connection is the order in the container */
+  const auto main_conn_cand = std::find_if (conn_vec.cbegin (), conn_vec.cend (),
+			      [] (const auto &conn)
   {
-    return conn_uptr.get () == conn;
+    return conn->is_connected ();
   });
 
-  if (conn_it == conn_vec.end ())
+  if (main_conn_cand == conn_vec.cend())
     {
-      return; // the connection is already cleared in disconnect_all_page_servers()
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_CONN_NO_PAGE_SERVER_AVAILABLE, 0);
+      return ER_CONN_NO_PAGE_SERVER_AVAILABLE;
     }
 
-  if (conn_it == conn_vec.begin ())
+  auto ulock = std::unique_lock<std::shared_mutex> { m_main_conn_mtx };
```

Comment:
Question:
Can consecutive push_request() or send_receive() calls make a starvation state?

Reply:
All they want to reset the main connection in the end, and it will be done. So, I don't think the starvation for resetting the main connection happens.

Reply:
By "starvation state" I understand that consecutive/continuous `shared_lock`s on the `m_main_conn_mtx` prevent resetting the main connection.
Yes, but very unlikely.
By comparison to the push/send methods, this one is a veeery rare event.

If the case, then we can use a `std::timed_mutex` and implement a loop here in `reset_main_connection`. This will, at least, allow us to log that starvation has happened.

Reply:
Oh,, true. It can be. Hmm..

Reply:
@hornetmj  I think we can leave it as it is now even without an issue to track this. It would be very rare, and we can easily notice it if the problem arises since every request keeps failing. What do you think?